[GeoMechanicsApplication] Implement hardening process for Mohr-Coulomb, based on the modified class structure

mnabideltares opened the pull request:
**📝 Description**
The code is copied and modified to match the recently implemented structure for Mohr-Coulomb

**🆕 Changelog**
- Added hardening
- Registered the coefficients, as arrays to account for different hardening types
- Added unit tests

mnabideltares committed:
set hardening with the recent structure

mnabideltares committed:
Fixes, and clang format

mnabideltares committed:
restructured and added unit tests

mnabideltares committed:
Merge remote-tracking branch 'origin/master' into geo/13826-use-new-structure-for-hardening

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -112,34 +120,55 @@ bool CoulombWithTensionCutOffImpl::IsAdmissibleSigmaTau(const Vector& rTrialSigm
+    int          counter     = 0;
+    while (error > 1.0e-8) {
+        const auto apex = CalculateApex(mCoulombYieldSurface.GetFrictionAngleInRadians(),
+                                        mCoulombYieldSurface.GetCohesion());
```
Since the location of the apex may now depend on $`\kappa`$, it may be wise to now make `CalculateApex` a `const` member function of class `CoulombYieldSurface`. In that case, we no longer need to pass the friction angle $`\phi`$ and cohesion $`c`$, since the class can retrieve those values by calling its members `GetFrictionAngleInRadians` and `GetCohesion` from the implementation of member `CalculateApex`. The code here would then look as follows:
```suggestion
        const auto apex = mCoulombYieldSurface.CalculateApex();
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -154,4 +183,18 @@ void CoulombWithTensionCutOffImpl::load(Serializer& rSerializer)
     rSerializer.load("TensionCutOff", mTensionCutOff);
 }
 
+double CoulombWithTensionCutOffImpl::CalculateEquivalentPlasticStrain(const Vector& rSigmaTau,
```
As suggested before, we may want to move this member to the `CoulombYieldSurface` class.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -622,4 +623,56 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_CalculateMaterialResponse
+    properties.SetValue(GEO_COHESION, 10.0);
+    properties.SetValue(GEO_DILATANCY_ANGLE, 0.0);
+    properties.SetValue(GEO_TENSILE_STRENGTH, 10.0);
+
```
Why is there a blank line here? The next three lines also set certain material properties.

avdg81 commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -622,4 +623,56 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_CalculateMaterialResponse
+    auto cauchy_stress_vector = UblasUtilities::CreateVector({10.0, 0.0, -40.0, 0.0});
+    auto expected_cauchy_stress_vector =
+        UblasUtilities::CreateVector({4.69104576482, 0.0, -34.69104576482, 0.0});
+    constexpr double tolerance = 1.0e-8;
```
```suggestion
    constexpr auto tolerance = 1.0e-8;
```

avdg81 reviewed (commented):
Hi Mohamed,

Thank you very much for preparing this clear and well-scoped PR. It was very much what I had expected. I have a few suggestions about the iteration loop as well as moving some functions around, but all of them are relatively minor things. Well done!

mnabideltares committed:
Modifications based on review

mnabideltares committed:
Bug is fixed, unit tests are fixed

mnabideltares committed:
Added keywords to material parameters

mnabideltares commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.cpp:
```diff
@@ -29,6 +29,11 @@ CoulombYieldSurface::KappaDependentFunction MakeConstantFunction(double Value)
 
+CoulombYieldSurface::KappaDependentFunction MakeLinearFunction(double Value, double Coefficient)
+{
+    return [Value, Coefficient](double kappa) { return std::max(Value + Coefficient * kappa, 0.0); };
```
Cohesion, friction and dilatancy angles must always stay positive. In the case of negatice, they need to be truncated. It is intentially set it by this way.

mnabideltares commented on applications/GeoMechanicsApplication/tests/cpp_tests/custom_constitutive/test_mohr_coulomb_with_tension_cutoff.cpp:
```diff
@@ -622,4 +623,56 @@ KRATOS_TEST_CASE_IN_SUITE(MohrCoulombWithTensionCutOff_CalculateMaterialResponse
+    properties.SetValue(GEO_TENSILE_STRENGTH, 10.0);
+
+    properties.SetValue(GEO_FRICTION_ANGLE_FUNCTION_COEFFICIENTS, UblasUtilities::CreateVector({0.0}));
+    properties.SetValue(GEO_COHESION_FUNCTION_COEFFICIENTS, UblasUtilities::CreateVector({1.3301270189}));
```
yes, it is based on analytical solution.

avdg81 committed:
Made two member functions `const`

avdg81 committed:
Removed unused `#include`

avdg81 committed:
Made `CalculateCornerPoint` a member of class `CoulombWithTensionCutOffImpl`

avdg81 committed:
Made `IsStressAtTensionApexReturnZone` a member of class `CoulombWithTensionCutOffImpl`

avdg81 committed:
Made `IsStressAtTensionCutoffReturnZone` a member of class `CoulombWithTensionCutOffImpl`

avdg81 committed:
Made `IsStressAtCornerReturnZone` a member of class `CoulombWithTensionCutOffImpl`

avdg81 committed:
Made `ReturnStressAtTensionApexReturnZone` a member of class `CoulombWithTensionCutOffImpl`

avdg81 committed:
Made `ReturnStressAtTensionCutoffReturnZone` a member of class `CoulombWithTensionCutOffImpl`

avdg81 committed:
Made `ReturnStressAtRegularFailureZone` a member of class `CoulombWithTensionCutOffImpl`

avdg81 committed:
When calculating the equivalent plastic strain, calculate the plastic multiplier rather than passing it through a function parameter

avdg81 committed:
Got rid of a local variable

avdg81 committed:
Rewrote the calculation of the equivalent plastic strain increment

It now uses proper mathematical structures rather than working things out at the component level.

avdg81 committed:
Make a truly linear function when you ask for it

avdg81 committed:
Moved the plastic tolerance and maximum number of iterations from the Coulomb yield surface to the Mohr-Coulomb implementation class

avdg81 committed:
Renamed two Geo variables and changed the type of one of them

mnabideltares committed:
Added check for material properties in Coulomb yield surface

mnabideltares committed:
Fixes

mnabideltares committed:
resetting of kappa is applied

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_with_tension_cut_off_impl.cpp:
```diff
@@ -104,51 +46,106 @@ bool CoulombWithTensionCutOffImpl::IsAdmissibleSigmaTau(const Vector& rTrialSigm
+{
+    // Check whether the tension cut-off lies beyond the apex
+    auto result                 = Vector{ZeroVector(2)};
+    result[0]                   = mCoulombYieldSurface.CalculateApex();
```
We can write this more compact (and perhaps slightly clearer) in the following way:
```suggestion
    auto result                 = UblasUtilities::CreateVector({mCoulombYieldSurface.CalculateApex(), 0.0});
```

avdg81 commented on applications/GeoMechanicsApplication/custom_constitutive/coulomb_yield_surface.cpp:
```diff
@@ -193,24 +176,61 @@ void CoulombYieldSurface::InitializeKappaDependentFunctions()
+           CalculatePlasticMultiplier(rSigmaTau, DerivativeOfFlowFunction(rSigmaTau, AveragingType));
+}
+
+void CoulombYieldSurface::CheckMaterialProperties()
```
When member functions can be made `const`, please do so:
```suggestion
void CoulombYieldSurface::CheckMaterialProperties() const
```

avdg81 committed:
Some minor suggestions

- Simplified some code by using a utility function that creates vectors.
- Made member `CheckMaterialProperties` `const`.
- Extracted a private member function that checks hardening coefficients (to reduce duplication).

avdg81 committed:
Adopt a different mechanism to restore kappa when needed

avdg81 reviewed (commented):
Hi Mohamed,
Thanks a lot for processing the review suggestions. After reviewing your latest changes, I have a few more (minor) suggestions. I believe we're almost there. Thanks for all the work that you have done.

mnabideltares committed:
Added unittests

mnabideltares committed:
Fixed unittests
